// @storybook/addon-ondevice-controls/dist/components/RadioSelect.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { styled } from '@storybook/react-native-theming';
const RadioContainer = styled.View(({ isInline }) => ({
    flexDirection: isInline ? 'row' : 'column',
    alignItems: isInline ? 'center' : 'flex-start',
    flexWrap: 'wrap',
    gap: 10,
}));
const RadioTouchable = styled.TouchableOpacity(() => ({
    alignItems: 'center',
    flexDirection: 'row',
}));
const RadioCircle = styled.View(({ theme }) => ({
    width: 16,
    height: 16,
    borderWidth: 1,
    borderColor: theme.appBorderColor,
    borderRadius: 8,
    backgroundColor: theme.background.content,
    alignItems: 'center',
    justifyContent: 'center',
}));
const RadioInnerCircle = styled.View(({ theme, selected }) => ({
    width: 8,
    height: 8,
    borderRadius: 4,
    backgroundColor: selected ? theme.color.positive : theme.background.content,
}));
const RadioLabel = styled.Text(({ theme }) => ({
    fontSize: theme.typography.size.s1,
    paddingStart: 10,
    color: theme.color.defaultText,
}));
const RadioSelect = ({ data = [], value = '', onChange, isInline }) => {
    return (_jsx(RadioContainer, { isInline: isInline, children: data.map((item) => (_jsxs(RadioTouchable, { activeOpacity: 0.7, onPress: () => {
                onChange(item);
            }, children: [_jsx(RadioCircle, { children: _jsx(RadioInnerCircle, { selected: value === item.key }) }), _jsx(RadioLabel, { children: item.label })] }, item.label))) }));
};
export default RadioSelect;
