// @storybook/addon-ondevice-controls/dist/types/Color.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from 'react';
import { View, TouchableWithoutFeedback, StyleSheet, Platform, Dimensions } from 'react-native';
import { styled, useTheme } from '@storybook/react-native-theming';
import { ColorPicker, fromHsv } from '../components/color-picker';
import { ModalPortal } from '../components/ModalPortal';
const TouchableContainer = styled.View(({ theme }) => ({
    width: 40,
    height: 40,
    borderWidth: 1,
    borderColor: theme.appBorderColor,
    borderRadius: 6,
    padding: 3,
    backgroundColor: theme.background.content,
}));
const Touchable = styled.TouchableOpacity(({ color }) => ({
    width: '100%',
    height: '100%',
    borderRadius: 4,
    backgroundColor: color,
}));
const ButtonTouchable = styled.TouchableOpacity(({ theme, primary }) => {
    return {
        backgroundColor: primary ? theme.color.secondary : theme.button.background,
        borderRadius: 6,
        borderWidth: 1,
        borderColor: primary ? theme.color.secondary : theme.button.border,
        paddingVertical: 6,
        paddingHorizontal: 20,
        justifyContent: 'center',
        alignItems: 'center',
    };
});
const ButtonText = styled.Text(({ theme, primary }) => {
    return {
        color: primary ? theme.color.inverseText : theme.color.defaultText,
        fontSize: theme.typography.size.s2,
        fontWeight: theme.typography.weight.bold,
    };
});
const ColorType = ({ arg, onChange = (value) => value }) => {
    const [displayColorPicker, setDisplayColorPicker] = useState(false);
    const [currentColor, setCurrentColor] = useState(null);
    const theme = useTheme();
    const openColorPicker = () => {
        setDisplayColorPicker(true);
    };
    const closeColorPicker = () => {
        setDisplayColorPicker(false);
    };
    const onChangeColor = (color) => {
        onChange(fromHsv(color));
    };
    if (Platform.OS === 'web') {
        return (_jsx("input", { type: "color", value: arg.value, onChange: (event) => onChange(event.target.value), style: {
                width: 40,
                height: 40,
                borderWidth: 1,
                borderColor: theme.appBorderColor,
                borderRadius: 6,
                padding: 2,
                backgroundColor: theme.background.content,
            } }));
    }
    return (_jsxs(View, { children: [_jsx(TouchableContainer, { children: _jsx(Touchable, { color: arg.value, onPress: openColorPicker }) }), _jsxs(ModalPortal, { supportedOrientations: ['portrait', 'landscape'], transparent: true, visible: displayColorPicker, onRequestClose: closeColorPicker, children: [_jsx(TouchableWithoutFeedback, { onPress: closeColorPicker, children: _jsx(View, { style: styles.modalOverlay }) }), _jsx(View, { style: styles.centerContainer, children: _jsxs(InnerContainer, { pointerEvents: "box-none", children: [_jsx(ColorPicker, { onColorSelected: onChangeColor, onColorChange: (color) => setCurrentColor(color), defaultColor: arg.value, oldColor: arg.value, style: styles.picker }), _jsxs(View, { style: styles.actionContainer, children: [_jsx(ButtonTouchable, { onPress: closeColorPicker, children: _jsx(ButtonText, { children: "Cancel" }) }), _jsx(View, { style: { width: 12 } }), _jsx(ButtonTouchable, { primary: true, onPress: () => {
                                                onChangeColor(currentColor);
                                                closeColorPicker();
                                            }, children: _jsx(ButtonText, { primary: true, children: "Select" }) })] })] }) })] })] }));
};
const InnerContainer = styled.View(({ theme }) => ({
    backgroundColor: theme.background.content,
    borderWidth: 1,
    borderColor: theme.appBorderColor,
    borderRadius: 10,
    margin: 24,
    padding: 10,
    maxWidth: 350,
    height: 400,
    maxHeight: Dimensions.get('screen').height - 24 * 2,
    boxShadow: '0px 0px 16px 0px rgba(0, 0, 0, 0.2)',
    elevation: 10,
}));
const styles = StyleSheet.create({
    actionContainer: {
        paddingTop: 8,
        alignItems: 'flex-end',
        flexDirection: 'row',
        justifyContent: 'flex-end',
        width: '100%',
    },
    picker: { flex: 1 },
    centerContainer: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    modalOverlay: {
        ...StyleSheet.absoluteFillObject,
        backgroundColor: 'rgba(0,0,0,0.3)',
    },
});
ColorType.serialize = (value) => value;
ColorType.deserialize = (value) => value;
export default ColorType;
